Kernel/FileSystem/FUSE/Inode.cpp

```diff
@@ -0,0 +1,309 @@
+        char* dirents = bit_cast<char*>(response->data() + sizeof(fuse_out_header));
+        u32 total_size = header->len - sizeof(fuse_out_header);
+        u32 offset = 0;
+        while (offset < total_size) {
+            fuse_dirent* dirent = bit_cast<fuse_dirent*>(dirents + offset);
+            if (dirent->ino == 0)
+                break;
+
+            if (dirent->namelen > NAME_MAX)
+                return Error::from_errno(EIO);
+
+            TRY(callback({ { dirent->name, dirent->namelen }, { fsid(), dirent->ino }, to_underlying(ram_backed_file_type_from_mode(dirent->type << 12)) }));
+            offset += get_dirent_entry_length_padded(dirent->namelen);
+        }
+        payload.offset = offset;
+    }
+
+    TRY(try_release(id, true));
+    return {};
+}
+
+ErrorOr<NonnullRefPtr<Inode>> FUSEInode::lookup(StringView name)
+{
+    auto name_buffer = TRY(KBuffer::try_create_with_size("FUSE: Lookup name string"sv, name.length() + 1));
+    memset(name_buffer->data(), 0, name_buffer->size());
+    memcpy(name_buffer->data(), name.characters_without_null_termination(), name.length());
+
+    auto response = TRY(fs().m_connection->send_request_and_wait_for_a_reply(FUSE_LOOKUP, identifier().index().value(), name_buffer->bytes()));
+
+    fuse_out_header* header = bit_cast<fuse_out_header*>(response->data());
+    if (header->error)
+        return Error::from_errno(-header->error);
+
+    fuse_entry_out* entry = bit_cast<fuse_entry_out*>(response->data() + sizeof(fuse_out_header));
+
+    InodeMetadata metadata {};
+
+    metadata.inode = { fsid(), entry->nodeid };
+    metadata.mode = entry->attr.mode;
+    metadata.link_count = entry->attr.nlink;
+    metadata.size = entry->attr.size;
+
+    return TRY(adopt_nonnull_ref_or_enomem(new (nothrow) FUSEInode(fs(), entry->nodeid, metadata)));
+}
+
+ErrorOr<void> FUSEInode::flush_metadata()
+{
+    return {};
+}
+
+ErrorOr<void> FUSEInode::add_child(Inode&, StringView, mode_t)
+{
+    return ENOTIMPL;
+}
+
+ErrorOr<NonnullRefPtr<Inode>> FUSEInode::create_child(StringView, mode_t, dev_t, UserID, GroupID)
+{
+    return ENOTIMPL;
+}
+
+ErrorOr<void> FUSEInode::remove_child(StringView)
+{
+    return ENOTIMPL;
+}
+
+ErrorOr<void> FUSEInode::replace_child(StringView, Inode&)
+{
+    return ENOTIMPL;
+}
+
+ErrorOr<void> FUSEInode::chmod(mode_t)
+{
+    return ENOTIMPL;
+}
+
+ErrorOr<void> FUSEInode::chown(UserID, GroupID)
+{
+    return ENOTIMPL;
+}
+
+ErrorOr<void> FUSEInode::truncate_locked(u64 new_size)
+{
+    VERIFY(m_inode_lock.is_locked());
+    VERIFY(!is_directory());
+
+    fuse_setattr_in setattr {};
+    setattr.valid = FATTR_SIZE;
+    setattr.size = new_size;
+
+    auto response = TRY(fs().m_connection->send_request_and_wait_for_a_reply(FUSE_SETATTR, identifier().index().value(), { &setattr, sizeof(setattr) }));
+
+    fuse_out_header* header = bit_cast<fuse_out_header*>(response->data());
+    if (header->error)
+        return Error::from_errno(-header->error);
```

Comment:
Something smelly about this code. We follow a raw pointer here, and we negate the return code. Can we at least `VERIFY` that the pointer is valid?

Reply:
Is there a better way than just casting it after the call site? maybe return `fuse_out_header`?